### PR TITLE
Split token instance fetcher

### LINKED ISSUE #7024: Token instance fetcher: extract realtime and catchup fetchers.
Problem:

Currently we have a single toke instance fetcher `Indexer.Fetcher.TokenInstance`. Initially, it collects token transfers in `Explorer.Chain.stream_unfetched_token_instances`, which have no created token instances. And then fetcher tries to collect metdata for that token instances and create rows in `token_instances` table. And if fetcher cannot get metadata for the token instance in several tries, fetching of this token instance will be retried after the full queue will be processed. The problem here is that, the number of such "bad" token instances maybe huge. And the queue will increase in time with new "bad" token instances coming. This will lead to the issue, that appearance of "good" new token instances (their metadata can be fetched without problems) is slowing down because of long queue of "failed" token instances.

Solution:

Divide token instance fetchers to 2 fetchers. The first one is like realtime fetcher, has small queue, which processing only new token instances. And if after several retries token instance cannot be fetched, it is marked to re-fetch and is sent to the queue of the 2nd fetcher which has long queue of "bad" token isntances to fetch. This approach would solve the problem of quick appearance of new token isntances. And on the other hand Blockscout still will try to process "bad" token instances until token developers fix `tokenURI` for token instance metadata.

### PR DESCRIPTION
**Pull request by nikitosing**
Close #7024 

## Changelog
Now 3 fetchers are responsible for fetching token instances:
- `Realtime` fetcher: 
  - processes token transfers which are not in `token_instances` table came from `Block.Fetcher`s
  - if token instance fetch successed then token instance added to `token_instances` table. If not then also added but with some `error`
- `Retry` fetcher: 
  - continiously refetches token instances: `SELECT * FROM token_instances WHERE error IS NOT NULL`
  - if token instance fetch successed then token instance updated. If not then do nothing
- `Sanitize` fetcher:
  - once and just after starting the server this fetcher process all the token instances which are not in token_instances table
  - if token instance fetch successed then token instance added to `token_instances` table. If not then also added but with some `error`

Remove:
- `INDEXER_TOKEN_INSTANCE_BATCH_SIZE` 
- `INDEXER_TOKEN_INSTANCE_CONCURRENCY`

Add: 
- `INDEXER_TOKEN_INSTANCE_RETRY_REFETCH_INTERVAL`
- `INDEXER_TOKEN_INSTANCE_RETRY_CONCURRENCY`
- `INDEXER_TOKEN_INSTANCE_REALTIME_CONCURRENCY`
- `INDEXER_TOKEN_INSTANCE_SANITIZE_CONCURRENCY`

https://github.com/blockscout/docs/pull/141

## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [x] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

**Review by vbaranov — COMMENTED** [review]
I would suggest changing the folders structure and naming of new fetcher's for consistency and reading convenience purposes. Here is the suggestion how to rename new fetchers:

Indexer.Fetcher.Realtime.TokenInstance -> Indexer.Fetcher.TokenInstance.Realtime
Indexer.Fetcher.Retry.TokenInstance -> Indexer.Fetcher.TokenInstance.Retry
Indexer.Fetcher.Sanitize.TokenInstance -> Indexer.Fetcher.TokenInstance.Sanitize
Indexer.Fetcher.TokenInstance -> Indexer.Fetcher.TokenInstance.Helper

And the folder structure like this:
apps/indexer/lib/indexer/fetcher/token_instance:
           - realtime.ex
           - retry.ex
           - sanitize.ex
           - helper.ex

File: apps/explorer/lib/explorer/token/instance_metadata_retriever.ex [hunk 1]
```diff
@@ -65,6 +65,10 @@ defmodule Explorer.Token.InstanceMetadataRetriever do
   # https://eips.ethereum.org/EIPS/eip-1155#metadata
   @erc1155_token_id_placeholder "{id}"
 
+  @max_error_length 100
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/token/instance_metadata_retriever.ex:68]: What is the rationale for 100? In the DB we have `character varying(255)` type for `error` column in `token_instances` table.

File: apps/indexer/lib/indexer/block/fetcher.ex [hunk 2]
```diff
@@ -25,7 +25,7 @@ defmodule Indexer.Block.Fetcher do
     ReplacedTransaction,
     Token,
     TokenBalance,
-    TokenInstance,
+    TokenInstance.Realtime,
```
Comment (vbaranov) [inline apps/indexer/lib/indexer/block/fetcher.ex:28]: I would suggest to make an alias for this fetcher for the sake of search improvement capabilities:
```
alias Indexer.Fetcher.TokenInstance.Realtime, as: TokenInstanceRealtime
```

File: apps/indexer/lib/indexer/fetcher/token_instance/helper.ex [hunk 3]
```diff
@@ -0,0 +1,47 @@
+defmodule Indexer.Fetcher.TokenInstance.Helper do
+  @moduledoc """
+    Common functions for Indexer.Fetcher.TokenInstance fetchers
+  """
+  alias Explorer.Chain
+  alias Explorer.Token.InstanceMetadataRetriever
+
+  def fetch_instance(token_contract_address_hash, token_id, is_retry?) do
```
Comment (vbaranov) [inline apps/indexer/lib/indexer/fetcher/token_instance/helper.ex:8]: Let's add spec for this function?

File: apps/indexer/lib/indexer/supervisor.ex [hunk 4]
```diff
@@ -24,7 +25,9 @@ defmodule Indexer.Supervisor do
     ReplacedTransaction,
     Token,
     TokenBalance,
-    TokenInstance,
+    TokenInstance.Realtime,
+    TokenInstance.Retry,
+    TokenInstance.Sanitize,
```
Comment (vbaranov) [inline apps/indexer/lib/indexer/supervisor.ex:30]: I would suggest to make an alias for these fetchers for the sake of search improvement capabilities:
```
alias Indexer.Fetcher.TokenInstance.Realtime, as: TokenInstanceRealtime
alias Indexer.Fetcher.TokenInstance.Retry, as: TokenInstance Retry
alias Indexer.Fetcher.TokenInstance.Sanitize, as: TokenInstanceSanitize
```

File: apps/indexer/lib/indexer/fetcher/token_instance/retry.ex [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+  @doc false
+  def child_spec([init_options, gen_server_options]) do
+    {state, mergeable_init_options} = Keyword.pop(init_options, :json_rpc_named_arguments)
+
+    unless state do
+      raise ArgumentError,
+            ":json_rpc_named_arguments must be provided to `#{__MODULE__}.child_spec " <>
+              "to allow for json_rpc calls when running."
+    end
```
Comment (sl1depengwyn) [inline apps/indexer/lib/indexer/fetcher/token_instance/retry.ex:27]: Same here

File: apps/indexer/lib/indexer/fetcher/token_instance/realtime.ex [hunk 6]
```diff
@@ -36,56 +35,19 @@ defmodule Indexer.Fetcher.TokenInstance do
...
   @impl BufferedTask
-  def init(initial_acc, reducer, _) do
-    {:ok, acc} =
-      Chain.stream_unfetched_token_instances(initial_acc, fn data, acc ->
-        reducer.(data, acc)
-      end)
-
-    acc
+  def init(_, _, _) do
+    {0, []}
```
Comment (sl1depengwyn) [inline apps/indexer/lib/indexer/fetcher/token_instance/realtime.ex:39]: Same in this module

File: apps/indexer/lib/indexer/supervisor.ex [hunk 7]
```diff
@@ -106,7 +109,15 @@ defmodule Indexer.Supervisor do
...
-        {TokenInstance.Supervisor,
+        {TokenInstanceRetry.Supervisor,
+         [
+           [json_rpc_named_arguments: json_rpc_named_arguments, memory_monitor: memory_monitor]
+         ]},
+        {TokenInstanceRealtime.Supervisor,
+         [
+           [json_rpc_named_arguments: json_rpc_named_arguments, memory_monitor: memory_monitor]
+         ]},
+        {TokenInstanceSanitize.Supervisor,
```
Comment (sl1depengwyn) [inline apps/indexer/lib/indexer/supervisor.ex:120]: Likewise we don't need to provide `json_rpc_named_arguments` here

**Review by sl1depengwyn — CHANGES_REQUESTED** [review]

File: apps/indexer/lib/indexer/fetcher/token_instance/sanitize.ex [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+  @default_max_concurrency 10
+  @doc false
+  def child_spec([init_options, gen_server_options]) do
+    {state, mergeable_init_options} = Keyword.pop(init_options, :json_rpc_named_arguments)
+
+    unless state do
+      raise ArgumentError,
+            ":json_rpc_named_arguments must be provided to `#{__MODULE__}.child_spec " <>
+              "to allow for json_rpc calls when running."
+    end
```
[the remark this answers is not in the document]
Comment (nikitosing) [reply apps/indexer/lib/indexer/fetcher/token_instance/sanitize.ex]: Good point. Fixed!

File: apps/indexer/lib/indexer/fetcher/token_instance/retry.ex (hunk 5 above)
Comment (nikitosing) [reply apps/indexer/lib/indexer/fetcher/token_instance/retry.ex]: Fixed!

File: apps/indexer/lib/indexer/fetcher/token_instance/realtime.ex (hunk 6 above)
Comment (nikitosing) [reply apps/indexer/lib/indexer/fetcher/token_instance/realtime.ex:31]: Fixed!

File: apps/indexer/lib/indexer/supervisor.ex (hunk 7 above)
Comment (nikitosing) [reply apps/indexer/lib/indexer/supervisor.ex]: Fixed!

**Review by sl1depengwyn — APPROVED** [review]

**Review by vbaranov — APPROVED** [review]

**Pull request merged**